pkg/operation/botanist/addons.go

```diff
@@ -359,6 +362,65 @@ func (b *Botanist) generateCoreAddonsChart(ctx context.Context) (*chartrenderer.
 		}
 	}
 
+	var (
+		poolNameVersionToImage = make(workerPoolKubeProxyImages)
+		kubernetesVersion      = b.Shoot.GetInfo().Spec.Kubernetes.Version
+	)
+
+	for _, worker := range b.Shoot.GetInfo().Spec.Provider.Workers {
+		image, err := b.ImageVector.FindImage(charts.ImageNameKubeProxy, imagevector.RuntimeVersion(kubernetesVersion), imagevector.TargetVersion(kubernetesVersion))
+		if err != nil {
+			return nil, err
+		}
+
+		poolNameVersionToImage.Upsert(worker.Name, kubernetesVersion, image.String())
+	}
+
+	nodeList := &corev1.NodeList{}
+	if err := b.K8sShootClient.Client().List(ctx, nodeList); err != nil {
+		return nil, err
+	}
+
+	for _, node := range nodeList.Items {
+		poolName, ok1 := node.Labels[v1beta1constants.LabelWorkerPool]
+		kubernetesVersion, ok2 := node.Labels[v1beta1constants.LabelWorkerKubernetesVersion]
+		if !ok1 || !ok2 {
+			continue
+		}
+
+		image, err := b.ImageVector.FindImage(charts.ImageNameKubeProxy, imagevector.RuntimeVersion(kubernetesVersion), imagevector.TargetVersion(kubernetesVersion))
+		if err != nil {
+			return nil, err
+		}
+
+		poolNameVersionToImage.Upsert(poolName, kubernetesVersion, image.String())
+	}
+
+	var workerPools []map[string]string
+
+	// TODO(rfranzke): Delete this in a future version.
```

Comment:
Why do we need this at all? I assume, because there could be unlabelled nodes and we don't want to force a rollout to label them? For new shoots, it's a bit confusing to have this daemonset (scaled to 0), and its vpa and configmap sitting there unused, but I guess it can't be helped?

Reply:
Update: I tested this with an existing shoot created with an older version and the "versioned" kube-proxy daemon set (e.g. `kube-proxy-worker-test-v1.21.2`) quickly ended up running also on the old nodes after they got their labels, and no node rollout was needed for this to happen. This is by design, correct? If yes, again the above question - could we get rid of the unversioned daemon set already now?

Reply:
Right, we do it only for backwards-compatibility/migration. Without this, the old `kube-proxy` `DaemonSet` would be deleted immediately while it can take up to `5m` until the new version label is added to all nodes. This could cause issues for end-users/workload.
We discussed that we could delete this `DaemonSet` in the next reconciliation, but since this is migration coding that will be deleted in an upcoming Gardener version anyways, we decided to save the "trouble".

Reply:
OK, during my tests adding the new label happened almost immediately, I wasn't aware it could take up to 5m. Thanks for explaining!